go/parquet/internal/utils/bit_writer.go

```diff
@@ -74,18 +79,19 @@ type BitWriter struct {
 
 // NewBitWriter initializes a new bit writer to write to the passed in interface
 // using WriteAt to write the appropriate offsets and values.
-func NewBitWriter(w io.WriterAt) *BitWriter {
+func NewBitWriter(w WriterAtWithLen) *BitWriter {
 	return &BitWriter{wr: w}
 }
 
-// ReserveBytes reserves the next aligned nbytes, skipping them and returning
+// SkipBytes reserves the next aligned nbytes, skipping them and returning
 // the offset to use with WriteAt to write to those reserved bytes. Used for
 // RLE encoding to fill in the indicators after encoding.
-func (b *BitWriter) ReserveBytes(nbytes int) int {
+func (b *BitWriter) SkipBytes(nbytes int) (int, error) {
 	b.Flush(true)
 	ret := b.byteoffset
 	b.byteoffset += nbytes
-	return ret
+	b.wr.Reserve(b.byteoffset)
```

Comment:
```Go
// Reserve pre-allocates nbytes to ensure that the next write of that many bytes
// will not require another allocation.
func (b *PooledBufferWriter) Reserve(nbytes int) {
	if b.buf == nil {
		b.buf = bufferPool.Get().(*memory.Buffer)
	}

	newCap := utils.MaxInt(b.buf.Cap()+b.offset, 256)
	for newCap < b.pos+nbytes {
		newCap = bitutil.NextPowerOf2(newCap)
	}
	b.buf.Reserve(newCap)
}
```

Would it Reserve more bytes than expected here?

Reply:
The bitwriter *only* uses `WriteAt` and never calls the regular `Write`. As a result the internal `b.pos` never gets incremented, so `Reserve` has to always include the whole size (in this situation) because we never advance the `b.pos` inside the buffer and only ever use `WriteAt`

Reply:
Oh yeah, I mean that, would we use more memory than expected? Maybe we have:

```
LevelEncoder( holds PooledBufferWriter)
-> BitWriter ( holds PooledBufferWriter)
--> Call Reserve ( b.byteoffset)
```

I'm not sure would it harms...

Reply:
hmm, can you elaborate? I'm not understanding what you're suggesting here.

Reply:
```go
func (b *BitWriter) SkipBytes(nbytes int) (int, error) {
	b.Flush(true)
	ret := b.byteoffset
	b.byteoffset += nbytes
	b.wr.Reserve(b.byteoffset)
```

(1) Here, the `b.wr.Reserve(b.byteoffset)` will try to Reserve the `b.byteoffset` bytes, which is the sum bytes the `WriterAtWithLen` owned.

And `PooledBufferWrite` will regard `Reserve(b.byteoffset)` as "reserve  `b.byteoffset` more bytes", would it allocate much more memory than it exactly want?

Reply:
Because the `BitWriter` only ever calls `WriteAt`, the buffer's `pos` variable never updates. (Only `Write` updates the internal `pos`, i.e. position). When we call `Reserve` it verifies that `pos + nbytes < capacity`. So the `BitWriter` needs to always call `Reserve` with the full `b.byteoffset` in order to ensure that the bytes get reserved correctly. So it won't end up allocating more than we want to (aside from the fact that we always round up to the next power of two when allocating with reserve so that we don't have too many allocations)

Reply:
Yeah I know what you mean. Assume that we want we `SkipBytes(4)`, and `b.byteoffset = 1024`.

1. `b.byteoffset = 1024 + 4`
2. `b.wr.Reserve(1028)`
3. dispatch to `PooledBufferWriter.Reserve(1028)`
4. `newCap: at lease 1024 + 1028`

I just want to make sure that in (4), would it better be `newCap: 1028`?

Reply:
currently, 4 would be `newCap: at least 1028`, yes.

Reply:
LGTM now, I'm no problem then